src/Raven.Server/Documents/AbstractBackgroundWorkStorage.cs

```diff
@@ -93,8 +93,37 @@ public Queue<DocumentExpirationInfo> GetDocuments(BackgroundWorkParameters optio
                             if (cancellationToken.IsCancellationRequested)
                                 return toProcess;
 
-                            if (TryEnqueueItemToProcess(options, ref totalCount, multiIt.CurrentKey, ticksAsSlice, toProcess) == false)
-                                break;
+                            var clonedId = multiIt.CurrentKey.Clone(options.Context.Transaction.InnerTransaction.Allocator);
+                            DocumentExpirationInfo item;
+                            try
+                            {
+                                item = GetDocumentAndId(options, clonedId, ticksAsSlice);
+                            }
+                            catch (DocumentConflictException)
+                            {
+                                item = new DocumentExpirationInfo { Status = DocumentExpirationInfoStatus.Conflict };
+                            }
+
+                            switch (item.Status)
+                            {
+                                case DocumentExpirationInfoStatus.Process when ShouldHandleWorkOnCurrentNode(options.DatabaseTopology, options.NodeTag) == false:
+                                case DocumentExpirationInfoStatus.Skip when ShouldHandleWorkOnCurrentNode(options.DatabaseTopology, options.NodeTag) == false:
```

Comment:
Why are we checking `ShouldHandleWorkOnCurrentNode` on _each entry_?
That is global for the current state of the cluster, no?
What work should we be doing here?

Reply:
yes it global per each run of `GetDocuments()`, I changed the code to calculate it once, when we enter this method.